airbyte-integrations/connectors/source-mssql/src/testFixtures/java/io/airbyte/integrations/source/mssql/MsSQLTestDatabase.java

```diff
@@ -89,45 +168,87 @@ public MsSQLTestDatabase withAgentStopped() {
+    int actualRecordCount;
+    int tryCount = 0;
+    do {
+      LOGGER.info(formatLogLine("fetching the number of CDC records for {}.{}, instance {}"), schemaName, tableName, cdcInstanceName);
+      try {
+        actualRecordCount = query(ctx -> ctx.fetch(sql)).get(0).get(0, Integer.class);
+      } catch (SQLException | DataAccessException e) {
+        actualRecordCount = 0;
+      }
+      LOGGER.info(formatLogLine("Found {} CDC records for {}.{} in instance {}. Expecting {}. Trying again ({}/{}"), actualRecordCount, schemaName,
+          tableName, cdcInstanceName,
+          recordCount, tryCount, MAX_RETRIES);
+      if (actualRecordCount >= recordCount) {
+        LOGGER.info(formatLogLine("found {} records after {} tries!"), actualRecordCount, tryCount);
+        return self();
+      }
+      Thread.sleep(1_000);
+    } while (tryCount++ < MAX_RETRIES);
+    throw new RuntimeException(
+        "failed to find %d records after %s seconds. Only found %d!".formatted(recordCount, MAX_RETRIES, actualRecordCount));
+  }
+
   public MsSQLTestDatabase withShortenedCapturePollingInterval() {
-    return with("EXEC sys.sp_cdc_change_job @job_type = 'capture', @pollinginterval = %d;",
-        MssqlCdcTargetPosition.MAX_LSN_QUERY_DELAY_TEST.toSeconds());
+    synchronized (getContainer()) {
+      return with("EXEC sys.sp_cdc_change_job @job_type = 'capture', @pollinginterval = 1;");
+    }
   }
 
-  private void waitForAgentState(final boolean running) {
+  private void waitForAgentState(final boolean running) throws Exception {
     final String expectedValue = running ? "Running." : "Stopped.";
-    LOGGER.debug("Waiting for SQLServerAgent state to change to '{}'.", expectedValue);
+    LOGGER.info(formatLogLine("Waiting for SQLServerAgent state to change to '{}'."), expectedValue);
     for (int i = 0; i < MAX_RETRIES; i++) {
       try {
         final var r = query(ctx -> ctx.fetch("EXEC master.dbo.xp_servicecontrol 'QueryState', N'SQLServerAGENT';").get(0));
         if (expectedValue.equalsIgnoreCase(r.getValue(0).toString())) {
-          LOGGER.debug("SQLServerAgent state is '{}', as expected.", expectedValue);
+          LOGGER.info(formatLogLine("SQLServerAgent state is '{}', as expected."), expectedValue);
           return;
         }
-        LOGGER.debug("Retrying, SQLServerAgent state {} does not match expected '{}'.", r, expectedValue);
+        LOGGER.info(formatLogLine("Retrying, SQLServerAgent state {} does not match expected '{}'."), r, expectedValue);
       } catch (final SQLException e) {
-        LOGGER.debug("Retrying agent state query after catching exception {}.", e.getMessage());
-      }
-      try {
-        Thread.sleep(1_000); // Wait one second between retries.
-      } catch (final InterruptedException e) {
-        throw new RuntimeException(e);
+        LOGGER.info(formatLogLine("Retrying agent state query after catching exception {}."), e.getMessage());
       }
+      Thread.sleep(1_000);
```

Comment:
It might be worth introducing a `retryEverySecondUntil` function which takes a closure and retries as long as that closure returns a null or something